File: 2019-2021/2학년 2학기/네트워크 플로우/돌멩이 제거.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#define LEFT_MAX 501
#define RIGHT_MAX 501

std::vector<int> connected[LEFT_MAX];
bool visited[LEFT_MAX];
int matching[RIGHT_MAX];

int DFS(int start)
{
    if (visited[start])
        return 0;

    visited[start] = true;

    for (int opposite : connected[start])
    {
        if (matching[opposite] == 0 || DFS(matching[opposite]))
        {
            matching[opposite] = start;
            return 1;
        }
    }

    return 0;
}

int bipartite(int N)
{
    int result = 0;

    for (int i = 1; i <= N; ++i)
    {
        std::fill(visited, visited + N+1, false);
        result += DFS(i);
    }

    return result;
}

int main()
{
    int n, k;
    scanf("%d %d", &n, &k);

    for (int i = 0; i < k; ++i)
    {
        int r, c;
        scanf("%d %d", &r, &c);
        connected[r].push_back(c);
    }

    printf("%d", bipartite(n));

    return 0;
}

File: 2019-2021/2학년 2학기/네트워크 플로우/돌멩이 제거.js
const { match } = require("assert");

const LEFT_MAX = 501;
const RIGHT_MAX = 501;

const connected = Array.from(Array(LEFT_MAX), () => Array());
const visited = Array.from(Array(LEFT_MAX), () => false);
const matching = Array.from(Array(RIGHT_MAX), () => 0);

function DFS(start) {
    if (visited[start])
        return 0;
    
    visited[start] = true;

    for (let opposite of connected[start]) {
        if (matching[opposite] === 0 || DFS(matching[opposite])) {
            matching[opposite] = start;
            return 1;
        }
    }

    return 0;
}

function bipartite(n) {
    let result = 0;

    for (let i = 1; i <= n; ++i) {
        visited.fill(false, 0, n+1);
        result += DFS(i);
    }

    return result;
}

const fs = require("fs");

let lines = fs.readFileSync("/dev/stdin", "utf-8").split('\n');
let [n, k] = lines[0].split(" ").map(v => Number(v));
for (let i = 1; i <= k; ++i) {
    let [r, c] = lines[i].split(" ").map(v => Number(v));
    connected[r].push(c);
}

console.log(bipartite(n));